File: filethumbnails/js/js/scenario1.js
//// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF
//// ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO
//// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//// PARTICULAR PURPOSE.
////
//// Copyright (c) Microsoft Corporation. All rights reserved

(function () {
    "use strict";

    var page = WinJS.UI.Pages.define("/html/scenario1.html", {
        ready: function (element, options) {
            document.getElementById("picture-thumb-start").addEventListener("click", pickPhoto, false);
        }
    });

    function pickPhoto() {
        // Map thumbnail modes to HTML <select> options
        var modes = {
            "picturesGrid": Windows.Storage.FileProperties.ThumbnailMode.picturesView,
            "picturesList": Windows.Storage.FileProperties.ThumbnailMode.listView,
            "singlePicture": Windows.Storage.FileProperties.ThumbnailMode.singleItem,
        };
        var modeNames = {
            "picturesGrid": "ThumbnailMode.picturesView",
            "picturesList": "ThumbnailMode.listView",
            "singlePicture": "ThumbnailMode.singleItem",
        };

        // Get scenario options from HTML
        var modeSelector = document.getElementById("picture-thumb-mode");
        var modeSelected = modeSelector.options[modeSelector.selectedIndex].value,
            isFastSelected = document.getElementById("picture-thumb-fastThumbnail").checked;

        // Set API parameters
        var requestedSize = 200,
            thumbnailMode = modes[modeSelected],
            thumbnailOptions = Windows.Storage.FileProperties.ThumbnailOptions.useCurrentScale;

        if (isFastSelected) {
            thumbnailOptions |= Windows.Storage.FileProperties.ThumbnailOptions.returnOnlyIfCached;
        }

        // Clean output in case of repeat usage
        cleanOutput();

        var openpicker = new Windows.Storage.Pickers.FileOpenPicker();
        openpicker.fileTypeFilter.replaceAll([".jpg", ".png", ".bmp", ".gif", ".tif"]);
        openpicker.suggestedStartLocation = Windows.Storage.Pickers.PickerLocationId.picturesLibrary;
        openpicker.pickSingleFileAsync().done(function (file) {
            if (file) {
                file.getThumbnailAsync(thumbnailMode, requestedSize, thumbnailOptions).done(function (thumbnail) {
                    if (thumbnail) {
                        outputResult(file, thumbnail, modeNames[modeSelected], requestedSize);
                    } else if (isFastSelected) {
                        WinJS.log && WinJS.log(SdkSample.errors.noExif, "sample", "status");
                    } else {
                        WinJS.log && WinJS.log(SdkSample.errors.noThumbnail, "sample", "status");
                    }
                }, function (error) {
                    WinJS.log && WinJS.log(SdkSample.errors.fail, "sample", "status");
                });
            } else {
                WinJS.log && WinJS.log(SdkSample.errors.cancel, "sample", "status");
            }
        });
    }

    function outputResult(item, thumbnailImage, thumbnailMode, requestedSize) {
        document.getElementById("picture-thumb-imageHolder").src = URL.createObjectURL(thumbnailImage, { oneTimeOnly: true });
        // Close the thumbnail stream once the image is loaded
        document.getElementById("picture-thumb-imageHolder").onload = function () {
            thumbnailImage.close();
        };
        document.getElementById("picture-thumb-modeName").innerText = thumbnailMode;
        document.getElementById("picture-thumb-fileName").innerText = "File used: " + item.name;
        document.getElementById("picture-thumb-requestedSize").innerText = "Requested size: " + requestedSize;
        document.getElementById("picture-thumb-returnedSize").innerText = "Returned size: " + thumbnailImage.originalWidth + "x" + thumbnailImage.originalHeight;
    }

    function cleanOutput() {
        WinJS.log && WinJS.log("", "sample", "status");
        document.getElementById("picture-thumb-imageHolder").src = "/images/placeholder-sdk.png";
        document.getElementById("picture-thumb-modeName").innerText = "";
        document.getElementById("picture-thumb-fileName").innerText = "";
        document.getElementById("picture-thumb-requestedSize").innerText = "";
        document.getElementById("picture-thumb-returnedSize").innerText = "";
    }
})();

File: Samples/FileThumbnails/cppwinrt/Scenario1.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "Scenario1.h"
#include "Scenario1.g.cpp"
#include "SampleConfiguration.h"

using namespace winrt;
using namespace Windows::Foundation;
using namespace Windows::Storage;
using namespace Windows::Storage::FileProperties;
using namespace Windows::Storage::Pickers;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Controls;
using namespace Windows::UI::Xaml::Navigation;

namespace winrt::SDKTemplate::implementation
{
    Scenario1::Scenario1()
    {
        InitializeComponent();
    }

    void Scenario1::OnNavigatedTo(NavigationEventArgs const&)
    {
        Helpers::ResetOutput(rootPage, ThumbnailImage(), OutputTextBlock());
    }

    fire_and_forget Scenario1::GetThumbnailButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();

        Helpers::ResetOutput(rootPage, ThumbnailImage(), OutputTextBlock());

        // Pick a photo
        FileOpenPicker openPicker;
        openPicker.FileTypeFilter().ReplaceAll(FileExtensions::Image());
        StorageFile file = co_await openPicker.PickSingleFileAsync();
        if (file != nullptr)
        {
            hstring thumbnailModeName = ModeComboBox().SelectedItem().as<ComboBoxItem>().Name();
            ThumbnailMode thumbnailMode = ThumbnailMode::ListView;
            if (thumbnailModeName == L"ListView")
            {
                thumbnailMode = ThumbnailMode::ListView;
            }
            if (thumbnailModeName == L"PicturesView")
            {
                thumbnailMode = ThumbnailMode::PicturesView;
            }
            else if (thumbnailModeName == L"SingleItem")
            {
                thumbnailMode = ThumbnailMode::SingleItem;
            }

            bool fastThumbnail = FastThumbnailCheckBox().IsChecked().Value();
            ThumbnailOptions thumbnailOptions = ThumbnailOptions::UseCurrentScale;
            if (fastThumbnail)
            {
                thumbnailOptions |= ThumbnailOptions::ReturnOnlyIfCached;
            }

            const uint32_t size = 200;
            StorageItemThumbnail thumbnail = co_await file.GetThumbnailAsync(thumbnailMode, size, thumbnailOptions);
            if (thumbnail != nullptr)
            {
                Helpers::DisplayResult(ThumbnailImage(), OutputTextBlock(), thumbnailMode, size, file, thumbnail, false);
                thumbnail.Close();
            }
            else if (fastThumbnail)
            {
                rootPage.NotifyUser(Errors::NoExifThumbnail, NotifyType::StatusMessage);
            }
            else
            {
                rootPage.NotifyUser(Errors::NoThumbnail, NotifyType::StatusMessage);
            }
        }
        else
        {
            rootPage.NotifyUser(Errors::Cancel, NotifyType::StatusMessage);
        }
    }

}
